ConstructingObjectParser adapts ObjectParser for ctor args

nik9000 opened the pull request:
ObjectParser makes parsing XContent 95% easier. No more nested loops.
No more forgetting to use ParseField. Consistent handling for arrays.
Awesome. But ObjectParser doesn't support building things objects whose
constructor arguments are mixed in with the rest of its properties.
Enter ConstructingObjectParser! ConstructingObjectParser queues up
fields until all of the constructor arguments have been parsed and
then sets them on the target object.

Closes #17352

nik9000 commented:
I labeled this discuss rather than review because it screws with error messages a bit too much. I can fix that if it integrate it more tightly into ObjectParser - either by hacking ObjectParser to expose something funky for ObjectSpooler, or, probably better, by extracting 90% of ObjectParser into an abstract base class and making ObjectSpooler inherit from that. In fact, I might just do that right now and see.

nik9000 commented:
OK - I was able to really really improve this. No more broken error messaging and cleaner syntax! I did it by pulling a big chunk of ObjectParser out into a superclass which both ObjectParser and ObjectSpooler extend. Then I modified it's implementation to ally ObjectSpooler to wrap all the consumers cleanly. Not all of ObjectParser made it to the interface because it isn't compatible, but most of it is.

I also sped it up a bit, removing a hash lookup from constructor arguments and removing the hash required.

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java:
```diff
@@ -207,66 +162,23 @@ public Value apply(XContentParser parser, Context context) {
         }
     }
 
-    public void declareStringArray(BiConsumer<Value, List<String>> consumer, ParseField field) {
```
Moved these to the superclass.

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java:
```diff
@@ -64,6 +64,7 @@
         };
     }
 
+    private final Map<String, FieldParser> fieldParserMap = new HashMap<>();
```
Moved this up from the middle of the class so it could be next to the other fields.

nik9000 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectParser.java:
```diff
@@ -146,52 +147,6 @@ public Value parse(XContentParser parser, Value value, Context context) throws I
         return value;
     }
 
-    private void parseArray(XContentParser parser, FieldParser<Value> fieldParser, String currentFieldName, Value value, Context context)
```
Moved these down because they are private.

nik9000 commented on core/src/main/java/org/elasticsearch/search/highlight/HighlightBuilder.java:
```diff
@@ -21,18 +21,12 @@
 import org.apache.lucene.search.Query;
 import org.apache.lucene.search.vectorhighlight.SimpleBoundaryScanner;
-import org.elasticsearch.common.ParsingException;
-import org.elasticsearch.common.io.stream.StreamInput;
```
These are just a leftover but it feels silly to go add unused imports back to the file just to remove it from the diff.

nik9000 commented:
@colings86 can you review this? I'm not sure who the right person is but we talked about it earlier today so maybe it is you? Feel free to kick it back to me if not.

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+ * allocates <code>2 + 2 + 2 * param_count</code> objects per parse. If this overhead is too much for you then feel free to have
+ * ObjectParser parse a secondary object and have that one call the target object's constructor.
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
```
Do we need a separate class for dealing with this case? Should this functionality maybe just be added to `ObjectParser` so we just have one thing that does parsing?

If we do decide to keep this class I think `ObjectSpooler` is not a great name as it doesn't really describe what it does. Maybe `SpooledObjectParser` would be better but I actually don't think it should contain `Spooled` as its hard to naturally know that relates to dealing with constructor args.

colings86 replied:
If we did move this functionality into ObjectParser we should have a check that completely avoids the spooling if no constructor args are declared (this  might be as simple as trying to build the `targetObject` as soon as we start parsing).

nik9000 replied:
So I thought about just adding this directly to ObjectParser. I didn't want to do it that way because ObjectParser is complex enough as it is and this'd just add more complexity. As a separate class that delegates to ObjectParser it feels like I'm at least containing the complexity required to support constructor arguments in its own place. And, when it is contained, it doesn't look *that* bad.

I'm totally open to another name. I don't care. `SpoolingObjectParser` is better than `SpooledObjectParser` but I don't particularly like either name. `QueueingObjectParser` is even worse though.

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + 2 * param_count</code> objects per parse. If this overhead is too much for you then feel free to have
```
I assume the figure here for the worst case is `2 + 2 + (2 * param_count)` rather than `(2 + 2 + 2) * param_count` but could you add the (unnecessary) brackets here to make it explicit and obvious which one it is

colings86 replied:
Maybe as an improvement (not necessarily for this PR) we could spool only until we have satisfied the constructor args and then call the remaining setters without spooling the values? This would make the best case (all constructor args are parsed before any other options) allocate only 2 additional objects per constructor arg, but would leave the worst case (final constructor arg is the last thing to be parsed) the same. Just an idea, not sure if it would add too much complexity though.

colings86 replied:
Oh, I think you have already implemented my idea above?

nik9000 replied:
Yeah, it already does that. It doesn't really add that much complexity and feels well worth it.

I just wrote `2 + 2 + 2 * param_count` because it lines up with how I think of it. And it lines up with the sentence above - the first `2` is the same as the `2` in that sentence.

colings86 replied:
Ok, but could we add brackets above so its obvious whether its `2 + 2 + (2 * param_count)` or `(2 + 2 + 2) * param_count`?

nik9000 replied:
644f680

jasontedor replied:
Arithmetic precedence rules negate the need for the parentheses here.

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+    }
+
+    @Override
+    public <T> void declareField(BiConsumer<Value, T> consumer, ContextParser<Context, T> parser, ParseField parseField, ValueType type) {
```
Could we avoid the `instanceof` check here by overloading this method and having a:
```
public <T> void declareField(ConstructorArg consumer, ContextParser<Context, T> parser, ParseField parseField, ValueType type);
``` 
method as well, or maybe better:

```
public <T> void declareConstructorField(ContextParser<Context, T> parser, ParseField parseField, ValueType type);
```

since we would always expect the consumer to be `this.constructorArg()`.

colings86 replied:
The second option here would actually mean that we wouldn't need the `ConstructorArg` class or the `this.constructorArg()` method since its only a marker anyway and the `declareConstructorField()` method would then act as the same marker

nik9000 replied:
The trouble with this idea is that it is really nice to be able to use `declareString` and friends when declaring the constructor args. I've made it a bit simpler in a patch I just pushed but it still has the marker field.

I did think about building `ctor()` method that returned the inner `ObjectParser` and then using the `constructorArg` method to build the actual consumer. The trouble with that is that you don't get access to the name there so you can't build the list of names for good error reporting. So I stuck with the marker.

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+     * {@link #builder} and the order that missing arguments are reported to the user if any are missing. When all of these parameters are
+     * parsed from the {@linkplain XContentParser} the target object is immediately built.
+     */
+    public <FieldT> BiConsumer<Value, FieldT> constructorArg() {
```
I think this should return ConstructorArg rather than BiFunction?

nik9000 replied:
I like returning BiConsumer here because it keeps ConstructorArg entirely hidden - it is a rather dirty implementation detail and I'd prefer it not leak into other stuff in the code so that if we clean it up we have an easier time of it. BTW, I just pushed a patch that removes the whole class in favor of a singleton BiConsumer so there isn't even a class to return in that case!

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+            constructorArgNames.add(parseField);
+            objectParser.declareField((spool, v) -> spool.constructorArg(position, v), parser, parseField, type);
+        }
+        objectParser.declareField(spoolingConsumer(consumer, parseField), parser, parseField, type);
```
Is it right that in the case of a constructorArg we call `objectParser.declareField()` twice here?

nik9000 replied:
Nope! I'll fix it. I'm not sure why, but ObjectParser ignores the second call. I mean, I know *why* it does that but I don't know why that is a good idea. I'll fix it.

nik9000 replied:
95dd165

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+         */
+        private void constructorArg(int position, Object value) {
+            if (constructorArgs[position] != null) {
+                throw new IllegalArgumentException("Can't repeat constructor param");
```
This message isn't going to mean a lot to the end user who reads it. I think we need to change it to `Can't repeat param [PARAM_NAME]` and pass the param name (`ParseField#getPreferredName()`) or `ParseField` itself into this method.

nik9000 replied:
Will do.

nik9000 replied:
a6a2918

colings86 commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,247 @@
+        private void spool(Consumer<Value> spoolMe) {
+            assert targetObject == null: "Don't spool after the targetObject has been built! Just apply the consumer directly.";
+            if (spooled == null) {
+                spooled = new ArrayList<>();
```
Do we know how many things (worst case) we are going to spool at this point? Is it just the number of fields declared to the parser? If so, should we preinitialise the array list with that size so it doesn't need to resize during parsing?

nik9000 replied:
Yeah, we know the worst case is the maximum number of things. I wonder if we expect that to be worth it most of the time. Maybe. I'm happy to use the size, especially if it is lower than the JVM's default which looks like it is 10. I'll have a think about it.

colings86 commented:
@nik9000 I left some comments but it looks good so far. I have assumed that apart from extracting the super class and changing method and variable order (as you put in comments) you have not changed the `ObjectParser` class and so have not really reviewed it deeply. If that is not a good assumption let me know and I will review that class.

nik9000 commented:
> If that is not a good assumption let me know and I will review that class.

Sorry! That isn't actually true.

The order of stuff in that class was really wacky and I tried not to change it *that* much but my eyes hurt seeing fields buried in the object. Those aren't big changes though.

The biggest change to ObjectParser is that it now has two declareField methods. One is the original for backward compatibility/"special" things. The other is compatible with the changes to the superclass. What I did was extract the superclass and realize that ObjectParser wanted to parse the field and set it to the object all in one go - and that was ok if I didn't want good error reporting. But to get nice error reporting I had to give ObjectSpooler a good way to wrap the consumer. So I created that second declareField method that takes a consumer and a ContextParser. This separated the two phases so that ObjectSpooler could do the wrapping it needed and ObjectParser just immediately applies the consumer to the result of the parse.

Then I had to change all the declareString, declareInt, declareXXX methods that I could do use the new declareField method. I don't think this change incurs any additional overhead and I think it actually makes it easier to read in the long run, but it is more than just extracting the methods.

Also that means the ObjectSpooler, or whatever we call it, doesn't support all the stuff that ObjectParser does - the things that relied on the original declareField method: declareNamedObjects and declareValue. I don't *think* that is going to be a problem. Or, at least, I think if that is a problem we can fix it later.

jasontedor commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,301 @@
+ * constructor arguments are first. On this platform with the same microbenchmarks just creating the XContentParser is around 900
+ * microseconds and using {#linkplain ObjectParser} directly adds another 300 or so microseconds. In the best case
+ * {@linkplain ObjectSpooler} allocates two additional objects per parse compared to {#linkplain ObjectParser}. In the worst case it
+ * allocates <code>2 + 2 + (2 * param_count)</code> objects per parse, plus a few ArrayList resizes for very long lists. If this overhead
```
It looks like we know the number of constructor arguments up front, so we can just use an `Object[]` of the right size and completely avoid the resizing?

jasontedor commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,301 @@
+ * is too much for you then feel free to have ObjectParser parse a secondary object and have that one call the target object's constructor.
+ * </p>
+ */
+public final class ObjectSpooler<Value, Context> extends AbstractObjectParser<Value, Context> {
```
I don't like the name `ObjectSpooler`. My biggest issue with it is that it's completely missing the fact that this is a parser but also it doesn't give a big indication that this is for parsing objects that have required arguments. I don't have a good suggestion, but hopefully my reasons for not liking it will help come up with something better?

nik9000 replied:
@colings86 had the same objections. So did I, actually. I just couldn't think of anything better. `ObjectConstructor`? It is less bad maybe? `ConstructingParser`?

jasontedor replied:
`ObjectWithRequiredFieldsParser`?

nik9000 replied:
`ConstructingObjectParser`.

jasontedor commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,301 @@
+import static java.lang.Math.min;
+
+/**
+ * Like {@link ObjectParser} but works with objects that have constructors who's arguments are mixed in with its other settings. Queries are
```
Nit: `who's` -> `whose`

nik9000 replied:
Damn English. Whose idea was this anyway?

jasontedor commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -0,0 +1,301 @@
+     * to spool lots of fields. They'll either not send that many fields or send the required fields before they get to this many fields.
+     * The actual value is fairly arbitrary but feels like "less than lots but plenty more spooling than we expect on average."
+     */
+    private static final int MAX_INITIAL_SIZE_OF_SPOOLED_FIELDS = 6;
```
I think this is an unneeded optimization, and I'd rather just avoid the resize reallocations?

nik9000 replied:
It probably isn't needed, yeah. We have a few things that have a dozen fields or so but most would be using ObjectParser anyway. I'll remove it.

nik9000 replied:
73e2a54

jasontedor commented on core/src/test/java/org/elasticsearch/common/xcontent/ObjectSpoolerTests.java:
```diff
@@ -0,0 +1,271 @@
+     * Tests the non-constructor fields are only set on time.
+     */
+    public void testCalledOneTime() throws IOException {
+        class CalledOneTime {
```
I think this should have a constructor argument to receive the value of the parsed field "yeah".

nik9000 replied:
376154c

jasontedor commented on core/src/test/java/org/elasticsearch/common/xcontent/ObjectSpoolerTests.java:
```diff
@@ -0,0 +1,271 @@
+              + "  \"yeah\": \"!\",\n"
+              + "  \"foo\": \"foo\"\n"
+              + "}");
+        spooler.apply(parser, null);
```
Can we capture the returned value and assert that `fooSet` is `true`?

nik9000 replied:
376154c

jasontedor commented on core/src/test/java/org/elasticsearch/common/xcontent/ObjectSpoolerTests.java:
```diff
@@ -0,0 +1,271 @@
+              + "  \"foo\": \"foo\",\n"
+              + "  \"yeah\": \"!\"\n"
+              + "}");
+        spooler.apply(parser, null);
```
Can we capture the returned value and assert that `fooSet` is `true`?

nik9000 replied:
376154c

jasontedor commented on core/src/test/java/org/elasticsearch/common/xcontent/ObjectSpoolerTests.java:
```diff
@@ -0,0 +1,271 @@
+     */
+    public void testRandomOrder() throws Exception {
+        int times = 100;
+        for (int i = 0; i < times; i++) {
```
Can we remove the poor-man's repeat?

nik9000 replied:
This is pretty common in our tests when we're pretty sure that doing it one time isn't going to get us useful coverage and the thing we do is fast. AbstractQueryTestCase does this and I had been working with it for a while before I did this.

I'm ok with removing it but I think we should maybe open up an issue about removing them all/unbanning repeat/something else?

jasontedor replied:
I think we get enough coverage from the CI?

The larger problem that I have with the poor man's repeat is that there are subtle issues with non-reproducible iterations. An example of this can be seen in #16208. If a non-deterministic number of random draws are taken on the `n`th iteration, then all further iterations are not reproducible.

jasontedor commented:
I think it looks great @nik9000, and solves a use-case I encountered earlier today! I left some comments.

nik9000 commented:
@jasontedor I believe I've addressed your comments. This will probably need some special love when I rebase it - specifically you can't use Void as a context any more.

jasontedor commented on core/src/main/java/org/elasticsearch/common/xcontent/ObjectSpooler.java:
```diff
@@ -225,29 +217,35 @@ private void constructorArg(int position, ParseField parseField, Object value) {
          * Spool up a consumer that we'll call once the targetObject is built. If targetObject has been built just apply the consumer
          * immediately.
          */
+        @SuppressWarnings("unchecked")
```
Can we do something like

```diff
diff --git a/core/src/main/java/org/elasticsearch/common/xcontent/ConstructingObjectParser.java b/core/src/main/java/org/elasticsearch/common/xcontent/ConstructingObjectParser.java
index b8fe09a..8d6b3ca 100644
--- a/core/src/main/java/org/elasticsearch/common/xcontent/ConstructingObjectParser.java
+++ b/core/src/main/java/org/elasticsearch/common/xcontent/ConstructingObjectParser.java
@@ -244,11 +244,12 @@ public final class ConstructingObjectParser<Value, Context> extends AbstractObje
          * Queue a consumer that we'll call once the targetObject is built. If targetObject has been built this will fail because the caller
          * should have just applied the consumer immediately.
          */
-        @SuppressWarnings("unchecked")
+
         private void queue(Consumer<Value> queueMe) {
             assert targetObject == null: "Don't queue after the targetObject has been built! Just apply the consumer directly.";
             if (queuedFields == null) {
-                this.queuedFields = new Consumer[numberOfFields];
+                @SuppressWarnings("unchecked") Consumer<Value>[] queuedFields = new Consumer[numberOfFields];
+                this.queuedFields = queuedFields;
             }
             queuedFields[queuedFieldsCount] = queueMe;
             queuedFieldsCount++;
```

so that the suppression applies to a single line instead of the whole method?

nik9000 replied:
That is what I did at first but I figured the method was so small it wan't worth the funky syntax dance. I can do it if you'd prefer though.

jasontedor replied:
> I can do it if you'd prefer though.

In general that is my preference. I've seen innocent method-applied suppressions end up suppressing something that should not have been suppressed that was later added to the code and fell under the existing suppression.

nik9000 replied:
Done. I also removed the loop in the test.

jasontedor commented:
> I believe I've addressed your comments.

@nik9000 I've reviewed it again. I left one more code-review comment and we need to wrap up the discussion around the poor man's repeat. I don't like the name `ConstructingObjectParser` but I can't come up with anything better.

jasontedor commented:
LGTM.
